hcal strip filter

aminnj opened the pull request:
Hcal problematic cell length filter added to BeamHaloSummary. Stored vector of problematic strips analogously to "PhiWedges" inside of HcalHaloData class.
- A description of the filter (+ motivation), along with application in Data/MC, is in the pair of slides in https://indico.cern.ch/event/438533/. 
- This adds ~130 bytes/evt to the reco BeamHaloSummary object
- As for CPU time impact, the main operation is just a single loop through all the calotowers. If this still requires timing information, I will run igprof and report on the results here

aminnj committed:
working filter

cmsbuild commented:
A new Pull Request was created by @aminnj (Nick Amin) for CMSSW_7_6_X.

hcal strip filter

It involves the following packages:

DataFormats/METReco
RecoMET/METAlgorithms
RecoMET/METFilters
RecoMET/METProducers


@cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please review it and eventually sign? Thanks.
@rappoccio, @ahinzmann, @mmarionncern, @jdolen, @nhanvtran, @schoef, @mariadalfonso, @TaiSakuma this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

slava77 commented:
@aminnj please provide more details:
- expected changes in RECO/AOD outputs or in MINIAOD outputs
- a link to slides in a JME (sub)group and maybe a plot or two to support the change

@schoef @mmarionncern please confirm the changes are as desired

aminnj commented:
Hi Slava,

- A description of the filter (+ motivation), along with application in Data/MC, is in the pair of slides in https://indico.cern.ch/event/438533/. 
- This adds ~130 bytes/evt to the reco BeamHaloSummary object
- As for CPU time impact, the main operation is just a single loop through all the calotowers. If this still requires timing information, I will run igprof and report on the results here

slava77 commented:
Hi Nick,
please move these bullets to the PR description (click pen-like "edit button" in the first block and edit the text there)

aminnj commented:
PR description updated

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: 589eba1d8a39ef06e1a6e01494819ab294126e49
 When I ran the RelVals I found an error in the following worklfows: 
 4.53 step3
<pre>runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODreHLT+HARVESTDreHLT/step3_RunPhoton2012B+RunPhoton2012B+HLTD+RECODreHLT+HARVESTDreHLT.log

</pre>
135.4 step3
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step3_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>
1306.0 step3
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step3_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>
1330.0 step3
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step3_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>
25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>
50202.0 step3
<pre>runTheMatrix-results/50202.0_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50/step3_TTbar_13+TTbar_13+DIGIUP15_PU50+RECOUP15_PU50+HARVESTUP15_PU50.log

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10934/7621/summary.html 

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/84ab4541df1baf9ffa428df9608db25cc1b74301
https://github.com/cms-sw/cmssw/commit/04aedc4bc6043e968db9a6b5f075f4f754e2a74e
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10934/7621/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10934/7621/git-merge-result

slava77 commented:
the configuration files couldn't parse
```
in _validateSequence
    raise RuntimeError("An entry in sequence "+label + ' has no label')
RuntimeError: An entry in sequence Flag_METFilters has no label
```

```runTheMatrix.py -l 4.53 ``` should be enough to reproduce

slava77 commented:
if I'm not mistaken, 
PhysicsTools/PatAlgos/python/slimming/metFilterPaths_cff.py should have the filter in the allMetFilterPaths and also listed as a separate path, not just as one filter in a master sequence metFilters

aminnj committed:
fix filter path for slimming

cmsbuild commented:
Pull request #10934 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

schoef commented:
@slava 
This filter is desired. To give some context, we've seen inefficiencies in the existing CSC halo filter and MET tails are populated by events with strip-like calorimetry noise. This filter identifies the pattern efficiently and we expect it to become standard quickly.

slava77 commented:
Hi Robert,

Please clarify if the implementation also looks good. Based on the tests
failed, it already looks like it's not following the MET filters setup
style.
On Aug 24, 2015 11:41 PM, "schoef" <notifications@github.com> wrote:

> @slava <https://github.com/slava>
> This filter is desired. To give some context, we've seen inefficiencies in
> the existing CSC halo filter and MET tails are populated by events with
> strip-like calorimetry noise. This filter identifies the pattern
> efficiently and we expect it to become standard quickly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10934#issuecomment-134471652>.
>

schoef commented:
@aminnj 
Please remove  the filter (until its studied) from the recommended filter here:
https://github.com/aminnj/cmssw/blob/hcal-cell-filter-76X/RecoMET/METFilters/python/metFilters_cff.py#L62

Should the FastSim not work, add the filter to the list here:
https://github.com/aminnj/cmssw/blob/hcal-cell-filter-76X/PhysicsTools/PatAlgos/python/slimming/metFilterPaths_cff.py#L38

@slava77 
It now looks correct. @aminnj Please test and give us the go ahead.

aminnj committed:
remove filter from main path and default to true in fastsim

aminnj commented:
I removed the filter from the main recommended filter path and defaulted the filter to True for fastsim. After doing so, the tests that failed above succeeded when I ran them manually. Can the bot confirm?

cmsbuild commented:
Pull request #10934 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

slava77 commented:
Nick, please check with ```runTheMatrix.py -l limited -i all ```

there will be more comments to the code (either from me or Carl later today or tomorrow)

slava77 commented on DataFormats/METReco/interface/BeamHaloSummary.h:
```diff
@@ -70,6 +73,12 @@ namespace reco {
     std::vector<int>& GetGlobaliPhiSuspects() { return GlobaliPhiSuspects ;}
     const std::vector<int>& GetGlobaliPhiSuspects() const { return GlobaliPhiSuspects ;}
 
+    std::vector<std::vector<std::pair<char, CaloTowerDetId> > >& GetProblematicStrips() { return ProblematicStrips ;}
```
Vectors of vectors are not good for persisted objects, we need to review these on a case-by-case basis.
You are saving a vector of "strip", where a "strip" is  a   vector of "CaloTowerDetId with a value" in this explicit vector-of-vectors form. A more reasonable solution would be to have a "CaloTowerIdsWithQuality" or a similar name that contains one "strip" and save a vector of these objects here in the BeamHaloSummary.
@Dr15Jones is it going to make things better from the I/O perspective?

I suggest to use uint8_t instead of a char
 (the rest of the code uses char for a byte-sized value though, the old parts could probably stay unchanged).

slava77 replied:
.. if you convert to CaloTowerIdsWithQuality, or just HaloTowerStrip for a more intuitive name, you can also put the hadEt value inside it to have one-stop access to the details /properties of the strip itself.

Dr15Jones replied:
A ```std::vector< [some type]>``` can be stored much more efficiently by ROOT (both in time and disk space) than a ```std::vector<std::vector<[some other type]>>```.

slava77 replied:
Instead of using a plain ```vector<vector``` a solution optimal for I/O is to linearize the 2D vector and save 1D vector + a vector of indices to point to start/end ranges of the data format.
I'm checking how large these things are on some data to see if this optimization is so significant.

slava77 commented on DataFormats/METReco/interface/BeamHaloSummary.h:
```diff
@@ -80,6 +89,9 @@ namespace reco {
     std::vector<int>  EcaliPhiSuspects;
     std::vector<int>  GlobaliPhiSuspects;
 
+    std::vector<std::vector<std::pair<char, CaloTowerDetId> > > ProblematicStrips;
```
please keep data members starting with a lower case.
I know, the rest of the file is different, but that's the old stuff that's in conflict with the cms style.

slava77 commented on DataFormats/METReco/interface/HcalHaloData.h:
```diff
@@ -20,9 +22,19 @@ namespace reco {
-    
+
+    // Return collection of problematic strips (pairs of # of problematic HCAL cells and CaloTowerDetId)
+    const std::vector<std::vector<std::pair<char, CaloTowerDetId> > >& GetProblematicStrips() const {return ProblematicStripCollection;}
```
same comments apply as for BeamHaloSummary

slava77 commented on DataFormats/METReco/src/classes_def.xml:
```diff
@@ -164,8 +164,8 @@
-  <class name="reco::HcalHaloData" ClassVersion="10">
-   <version ClassVersion="10" checksum="3928363167"/>
+  <class name="reco::HcalHaloData" ClassVersion="11">
+   <version ClassVersion="11" checksum="624180652"/>
```
old "ClassVersion-checksum" pairs should always be kept if there was data saved with them.
add ```<version ClassVersion="10" checksum="3928363167"/>``` right below what you have in L168

slava77 commented on DataFormats/METReco/src/classes_def.xml:
```diff
@@ -179,8 +179,8 @@
-  <class name="reco::BeamHaloSummary" ClassVersion="10">
-   <version ClassVersion="10" checksum="1514181401"/>
+  <class name="reco::BeamHaloSummary" ClassVersion="11">
+   <version ClassVersion="11" checksum="2109605555"/>
```
keep ``` <version ClassVersion="10" checksum="1514181401"/> ``` in the file

slava77 commented on PhysicsTools/PatAlgos/python/slimming/metFilterPaths_cff.py:
```diff
@@ -2,7 +2,7 @@
 ## We don't use "import *" because the cff contains some modules for which the C++ class doesn't exist
 ## and this triggers an error under unscheduled mode
-from RecoMET.METFilters.metFilters_cff import HBHENoiseFilterResultProducer, HBHENoiseFilter, HBHENoiseIsoFilter, CSCTightHaloFilter, hcalLaserEventFilter, EcalDeadCellTriggerPrimitiveFilter, eeBadScFilter, ecalLaserCorrFilter, EcalDeadCellBoundaryEnergyFilter, primaryVertexFilter 
+from RecoMET.METFilters.metFilters_cff import HBHENoiseFilterResultProducer, HBHENoiseFilter, HBHENoiseIsoFilter, CSCTightHaloFilter, HcalStripHaloFilter, hcalLaserEventFilter, EcalDeadCellTriggerPrimitiveFilter, eeBadScFilter, ecalLaserCorrFilter, EcalDeadCellBoundaryEnergyFilter, primaryVertexFilter 
```
please break this up into several import statements
The one-line import is becoming unreadable

slava77 commented on RecoMET/METAlgorithms/src/HcalHaloAlgo.cc:
```diff
@@ -119,6 +127,66 @@ HcalHaloData HcalHaloAlgo::Calculate(const CaloGeometry& TheCaloGeometry, edm::H
+
+  // Don't use HF.
+  int maxAbsIEta = 29;
+  std::vector<const CaloTower*> SortedCaloTowers;
```
please keep local variables names starting with lower case (CMS style); first letter capital is reserved for type names

slava77 commented on RecoMET/METAlgorithms/src/HcalHaloAlgo.cc:
```diff
@@ -119,6 +127,66 @@ HcalHaloData HcalHaloAlgo::Calculate(const CaloGeometry& TheCaloGeometry, edm::H
+  bool wasContiguous = true;
+  // Loop through and store a vector of pairs (problematicCells, DetId) for each contiguous strip we find
+  for(unsigned int i = 0; i < SortedCaloTowers.size(); i++) {
+    const CaloTower* tower = SortedCaloTowers.at(i);
```
``` sortedCaloTowers[i]; ```
.at() is too excessive considering you know already that i runs inside the range

slava77 commented on RecoMET/METAlgorithms/src/HcalHaloAlgo.cc:
```diff
@@ -119,6 +127,66 @@ HcalHaloData HcalHaloAlgo::Calculate(const CaloGeometry& TheCaloGeometry, edm::H
+
+    towerPair = std::make_pair((char)problematicCells, tower->id());
+
+    bool newIPhi = (tower->iphi()-1 == prevIPhi) || (i == 0);
```
shouldn't this be just ``` tower->iphi() != prevIPhi  ``` ?

slava77 commented on RecoMET/METFilters/plugins/HcalStripHaloFilter.cc:
```diff
@@ -0,0 +1,57 @@
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "DataFormats/METReco/interface/BeamHaloSummary.h"
+
+class HcalStripHaloFilter : public edm::EDFilter {
```
new legacy modules are not allowed. Module type should be thread-aware 
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkModuleTypes
edm::global::EDFilter would be a better choice.

slava77 commented on RecoMET/METFilters/plugins/HcalStripHaloFilter.cc:
```diff
@@ -0,0 +1,57 @@
+  iEvent.getByToken(beamHaloSummaryToken_ , beamHaloSummary);
+
+  bool pass = true;
+  std::vector<std::vector<std::pair<char, CaloTowerDetId> > > problematicStrips = beamHaloSummary->GetProblematicStrips();
```
you are not modifying the strips here, so use const ref
``` auto const& problematicStrips = beamHaloSummary->GetProblematicStrips();```

slava77 commented on RecoMET/METFilters/plugins/HcalStripHaloFilter.cc:
```diff
@@ -0,0 +1,57 @@
+  // std::vector<float> problematicStripsHadEt = beamHaloSummary->GetProblematicStripsHadEt();
+  for (unsigned int iStrip = 0; iStrip < problematicStrips.size(); iStrip++) {
+    int numContiguousCells = 0;
+    std::vector<std::pair<char, CaloTowerDetId> > problematicStrip = problematicStrips.at(iStrip);
```
don't use at(; also don't make a copy:
``` auto const& problematicStrip = problematicStrips[iStrip]; ```

slava77 commented on RecoMET/METProducers/src/BeamHaloSummaryProducer.cc:
```diff
@@ -210,6 +212,17 @@ void BeamHaloSummaryProducer::produce(Event& iEvent, const EventSetup& iSetup)
 
+
+  for( unsigned int i = 0 ; i < HcalData.GetProblematicStrips().size() ; i++ ) {
+    std::vector<std::pair<char, CaloTowerDetId> > ProblematicStrip = HcalData.GetProblematicStrips().at(i);
```
``` auto const& problematicStrip = HcalData.GetProblematicStrips()[i]; ```

aminnj committed:
use struct and change member capitalization

slava77 commented:
@cmsbuild please test

slava77 commented:
Nick, please clarify if you have checked that the updated code still behaves as you want it to.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #10934 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

aminnj commented:
Yes. This commit addresses your comments.
On Aug 28, 2015 8:43 AM, "Slava Krutelyov" <notifications@github.com> wrote:

> Nick, please clarify if you have checked that the updated code still
> behaves as you want it to.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/10934#issuecomment-135810180>.
>

cmsbuild commented:
-1 
Tested at: c4955b33627bf82b9d6ab0ac90cd3908c687ea27
 When I ran the RelVals I found an error in the following worklfows: 
 1000.0 step3
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step3_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10934/7724/summary.html

aminnj commented:
Will check

slava77 commented:
@cmsbuild please test

the error shows up with a segfault with not very descriptive stack trace fragments
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10934/7724/runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step3_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1 
Tested at: c4955b33627bf82b9d6ab0ac90cd3908c687ea27
 When I ran the RelVals I found an error in the following worklfows: 
 1000.0 step3
<pre>runTheMatrix-results/1000.0_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT/step3_RunMinBias2011A+RunMinBias2011A+TIER0+SKIMD+HARVESTDfst2+ALCASPLIT.log

</pre>

 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10934/7734/summary.html

aminnj commented:
Using gdb, I find this is related to freeing the memory associated with the struct I introduced. I'll fix it up and check again.

aminnj committed:
add class def and switch to uint_8

aminnj commented:
Segfault no longer occurs. It was due to object storage. Test 1000.0 works and  `runTheMatrix.py -l 4.22,4.53,5.1,8.0,9.0,25.0,101.0,135.4,140.53,1000.0,1001.0,1003.0,1306.0,1330.0,25202.0,50202.0` from above gives `16 15 14 11 5 1 tests passed, 0 0 0 0 0 0 failed`.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #10934 was updated. @cmsbuild, @cvuosalo, @vadler, @monttj, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 51b272c06f8d8a6ef83b4ff53d84c9a582d617cf
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10934/7878/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-10934/7878/summary.html

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/10934	51b272c
- changes in the code are OK now.
- jenkins test passed and comparisons with baseline don't show any relevant differences (as expected from the set of monitored quantities)
- checked locally in CMSSW_7_6_X_2015-08-30-2300 /test area sign582/ on run2 prompt reco of 200 events in 251721 DoubleMuon and DoubleEG
    - as expected, changes appear in recoBeamHaloSummary BeamHaloSummary and in recoHcalHaloData HcalHaloData values (based on the size on disk)
        - HcalHaloData is saved only in RECO tier
        - BeamHaloSummary is available in RECO and in AOD (not in MINIAOD). Its size in AOD increases from 104 bytes/evt to 120 bytes/event. (Curiously, in RECO data tier it's size changes only from 80 to 82 bytes/event in the DoubleMuon test; this is just size vs compression for small event count and full-split vs per-branch storage)
    - CPU use is essentially the same

slava77 commented:
@monttj please check if this PR is OK from AT side.

@davidlange6 this PR has backports to 74X/75X. The code can appropriately run only on RECO inputs and the computed values are relevant for data, not so much for MC.

monttj commented:
+1 OK from AT side

slava77 commented:
@monttj your +1 was not recognized as such
please send another one with just "+1" on a single line

slava77 commented:
@davidlange6 please merge: the only missing signature is from AT and it's missing only because of a typo

slava77 commented on DataFormats/METReco/src/classes_def.xml:
```diff
@@ -164,11 +164,15 @@
   </class>
   <class name="edm::Wrapper<reco::EcalHaloData>"/>
```
@Dr15Jones 
what do you think is missing in this file so that it will be seen in 1102, which is just a read of a file and then write with 
"fastCloning = False"
?

I was able to reproduce by merging this PR in 10934
(in my pre-signoff tests I missed to run the matrix locally)

Dr15Jones replied:
@slava77 I'm not sure what you are asking. What was the actual failure you see? Is this from the exception containing the message
"Incorrect size read for the container in recoBeamHaloSummary_BeamHaloSummary__reRECO.obj.problematicStrips"
?

Dr15Jones replied:
Looks like ```HaloTowerStrip``` doesn't have a statement that it should have its dictionary generated. There is only one for ```std::vector<HaloTowerStrip>```

slava77 replied:
yes, that's the message/error.

slava77 replied:
@aminnj Nick, please follow up (check in 76X and make a fix PR for 76X; and then update the 74X/75X PRs)

aminnj replied:
Hi Slava, I have added a line for HaloTowerStrip. It now passes `runTheMatrix.py -l 1102 -i all`:
https://github.com/cms-sw/cmssw/pull/11447

Similarly, I have updated 74X, 75X versions with the same exact changes
